soroban-env-host/src/events/internal.rs

```diff
@@ -141,15 +142,24 @@ pub(crate) struct InternalEventsBuffer {
 
 impl InternalEventsBuffer {
     // Records an InternalEvent
-    pub fn record(&mut self, e: InternalEvent, _budget: &Budget) -> Result<(), HostError> {
-        //TODO:Add metering for non-diagnostic events
+    pub fn record(&mut self, e: InternalEvent, budget: &Budget) -> Result<(), HostError> {
+        // Metering: we use the cost of instantiating a size=1 `Vec` as an estimate for the cost
+        // `Vec.push(event)`. Because the buffer length may be different on different instances
+        // due to diagnostic events and we need a deterministic cost across all instances,
+        // the cost needs to be amortized and buffer size-independent.
+        metered_clone::charge_container_bulk_init_with_elts::<
```

Comment:
Does this also charge for Diagnostic events?

Reply:
Oh you are right, good catch! 
Fixed.

Reply:
I've added a couple of tests.